examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3674 @@
+fn contract_wrap<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation, StateApiType = S>,
+    amount: Amount,
+) -> ContractResult<()> {
+    let (proxy_address, state_address) = get_protocol_addresses_from_implementation(host)?;
+
+    // Can be only called through the fallback function on the proxy.
+    only_proxy(proxy_address, ctx.sender())?;
+
+    // Check that contract is not paused.
+    when_not_paused(&state_address, host)?;
+
+    if amount == Amount::zero() {
+        return Ok(());
+    }
+
+    let input: WrapParamsWithSender = ctx.parameter_cursor().get()?;
+
+    // Get the sender who invoked this contract function.
+    let sender = input.sender;
+
+    let receive_address = input.params.to.address();
+
+    // Proxy holds CCD funds. CCD is sent to proxy.
+    host.invoke_contract_raw(
+        &proxy_address,
+        Parameter(&[]),
+        EntrypointName::new_unchecked("receiveCCD"),
+        amount,
+    )?;
+
+    // Update state.
+    host.invoke_contract(
+        &state_address,
+        &SetBalanceParams {
+            owner:  receive_address,
+            amount: amount.micro_ccd.into(),
+            update: Update::Add,
+        },
+        EntrypointName::new_unchecked("setBalance"),
+        Amount::zero(),
+    )?;
+
+    // Log the newly minted tokens.
+    host.invoke_contract(
+        &proxy_address,
+        &Cis2Event::Mint(MintEvent {
+            token_id: TOKEN_ID_WCCD,
+            amount:   ContractTokenAmount::from(amount.micro_ccd),
+            owner:    sender,
+        }),
+        EntrypointName::new_unchecked("logEvent"),
+        Amount::zero(),
+    )?;
+
+    // Only logs a transfer event if the receiver is not the sender.
+    // Only executes the `OnReceivingCis2` hook if the receiver is not the sender
+    // and the receiver is a contract.
+    if sender != receive_address {
+        // Log the transfer event.
+        host.invoke_contract(
+            &proxy_address,
+            &Cis2Event::Transfer(TransferEvent {
+                token_id: TOKEN_ID_WCCD,
+                amount:   ContractTokenAmount::from(amount.micro_ccd),
+                from:     sender,
+                to:       receive_address,
+            }),
+            EntrypointName::new_unchecked("logEvent"),
+            Amount::zero(),
+        )?;
+
+        // If the receiver is a contract: invoke the receive hook function.
+        if let Receiver::Contract(address, function) = input.params.to {
+            host.invoke_contract(
+                &address,
+                &OnReceivingCis2Params {
+                    token_id: TOKEN_ID_WCCD,
+                    amount:   ContractTokenAmount::from(amount.micro_ccd),
+                    from:     sender,
+                    data:     input.params.data,
+                },
+                function.as_entrypoint_name(),
+                Amount::zero(),
+            )?;
```

Comment:
I think we should move the invoking of receive hook to the proxy, since this acts as the token contract to the outside.
The smart contract receiving this, would see the implementation contract as the sender and might assume the sender is the token contract.

Reply:
moved to proxy